ext/cl_khr_tensor.asciidoc

```diff
@@ -0,0 +1,547 @@
+  i1 * tensor.shape[2] * tensor.shape[3] * ... * tensor.shape[N-1] +
+  ... +
+  i<N-2> * tensor.shape[i(N-1)] +
+  i<N-1>]
+----
+
+Where `iX` is a tensor coordinate index with inclusive range of `0..<shape[X]>`.
+
+// TODO: add clEnqueueCopyTensor
+
+// TODO: add clEnqueueFillTensor?
+
+// TODO: add command buffer variants for clEnqueue{copy,read,write}Tensor.
+
+
+==== Add New Buffer Property in Section 5.2.1
+
+[cols="2,1,2",stripes=odd]
+|===
+| CL_MEM_BIND_TO_TENSOR | cl_tensor a| Use the created buffer as
+storage for the given valid tensor. To succeed creating the buffer,
+the target tensor may not have storage already, must not have
+CL_TENSOR_COMMAND_BUFFER_TEMPORARY property set on and _size_ argument
+of the clCreateBufferWithProperties() must be zero.
+
+Size of the memory buffer is implementation-defined and it can be
+queried with clGetTensorInfo().
+
+Memory layout of the tensor in the created memory buffer is
+implementation-defined and opaque to the applications and it may
+change at unspecified points. Implementation may store auxiliary data
+in the memory buffer for the tensor. Therefore, writing data into the
+memory buffer directly using the cl_mem handle leads to undefined
+behavior.
+
+If the tensor is already bound to a buffer object,
+clCreateBufferWithProperties call returns CL_TENSOR_BOUND_TO_BUFFER
+error code.
+|===
+
+=== Sample Codes
+
+Helper functions used in the follow up tensor code samples:
+
+[source,c]
+----
+cl_kernel create_matmul_kernel(
+  cl_context ctx, std::span<cl_device_id> device_span,
+  cl_tensor lhs, cl_tensor rhs, cl_tensor out) {
+  // A hypothetical matmul kernel signature in pseudo OpenCL C for
+  // illustrative purposes:
+  //
+  //   kernel void matmul(
+  //     global read_only tensor_t,
+  //     global read_only tensor_t,
+  //     global write_only tensor_t);
+
+  cl_kernel matmul_kernel = /* Omitted. */;
+  clSetKernelArg(matmul_kernel, 0, sizeof(cl_tensor), &lhs);
+  clSetKernelArg(matmul_kernel, 1, sizeof(cl_tensor), &rhs);
+  clSetKernelArg(matmul_kernel, 2, sizeof(cl_tensor), &out);
+  return matmul_kernel;
+}
+
+cl_kernel create_matmul_kernel(
+  cl_context ctx, std::span<cl_device_id> device_span,
+  cl_tensor lhs, cl_tensor rhs, cl_tensor out) {
+  // A hypothetical add kernel signature in pseudo OpenCL C for illustrative
+  // purposes:
+  //
+  // kernel void add(
+  //     global read_only tensor_t,
+  //     global read_only tensor_t,
+  //     global write_only tensor_t);
+
+  cl_tensor add_kernel = /* Omitted. */;
+  clSetKernelArg(add_kernel, 0, sizeof(cl_tensor), &lhs);
+  clSetKernelArg(add_kernel, 1, sizeof(cl_tensor), &rhs);
+  clSetKernelArg(add_kernel, 2, sizeof(cl_tensor), &out);
+  return add_kernel;
+}
+----
+An example usage of tensors on a command queue:
+
+[source,c]
+----
+constexpr size_t b = 64, m = 100, n = 200, k = 50;
+
+cl_tensor in0 = clCreateTensor(ctx, nullptr, 3, {b, m, k}, CL_TENSOR_FLOAT, err);
```

Comment:
If we stick to C (like we should since the OpenCL spec is a C spec), we should use NULL instead of nullptr, right?